src/lib/yang/alarms.lua

```diff
@@ -348,11 +352,20 @@ end
 -- status change is added to the alarm.
 local function new_alarm (key, args)
    local ret = assert(alarm_list:retrieve(key, args), 'Not supported alarm')
+
+   local alt_resource = args.alt_resource or ret.alt_resource or {}
+   assert(type(alt_resource) == "table")
+
    local status = {
       time = format_date_as_iso_8601(),
       perceived_severity = args.perceived_severity or ret.perceived_severity,
       alarm_text = args.alarm_text or ret.alarm_text,
+      resource      = key.resource,
```

Comment:
An alarm-status-change is defined in the IETF YANG schema as (https://www.ietf.org/archive/id/draft-vallin-netmod-alarm-module-02.txt):

```
+--ro status-change* [time] {alarm-history}?
       |     |  +--ro time                  yang:date-and-time
       |     |  +--ro perceived-severity    severity-with-clear
       |     |  +--ro alarm-text            alarm-text
```

If you add extra information here, I'm not sure it will be shown in those tools that serialize/deserialize the data according to the schema (for instance, `alarms get`). In any case, this change doesn't respect the schema, so it's not correct.

Reply:
Okay - I think I understand now - I added these to the status item as, as far as I could tell, this is what is outputted by `alarms listen` - I was under the impression that the output of `alarms listen` was an `alarm-notification` but as you point out, it's currently a `status-change` item.

Which makes me think that there's missing functionality here - when a `status-change` occurs and is added to the list, an `alarm-notification` should be sent which contains the full information required. Does this sound correct? Is `alarms listen` *designed* to only send a `status-change` and if so, what's the correct way to work with it to retrieve the resource(s) affected?

Reply:
`alarm listen` is the right command for listening for notifications. `alarm get-state` is like `config state`, it returns state information of the lwAFTR instance but in this case it only reports alarms.

According to the schema, there are three types of notifications:
   * alarm-notification
   * alarm-inventory-changed
   * operator-action

An alarm-notification is sent in the following scenarios:
   * When an alarm is raised.
   * When an alarm is cleared.
   * When there's a status-change in the alarm (for example, because the alarm changes it severity).

Point is since you need a lwAFTR instance before running `alarms listen`, if your alarms are raised between the time you started the lwAFTR and run the listen command, you will miss those notifications. To test whether a notification is sent or not maybe what you could try is to fix the condition that raised your alarm (so it should get cleared and send a notification). I don't know if it's feasible to do that with the link_down alarm you added.

Another possibility would be to add a print in the `add_alarm_notification` and see if it's being executed (definitely not the best way to check if it's working).

As far as I recall, I tested `alarm-inventory-changed` and `operator-action` and checked they were working, but I can't tell now about `alarm-notification`. Maybe there's an issue going on there.

About the info that is going to print out, it would be all the info that is available in the alarm I think. There are fields that are specified in the schema but I didn't supported (such as `alt-resource` but others are `related-alarm`, `impacted-resource` or `root-cause-resource`). IMHO the alarms schema is overly complicated and with such a lot of fields for an alarm, encoding/decoding of alarms got more complicated too. That's why I initially decided to support only a limited amount of fields (the most fundamental). with the idea that if someone needs more fields in the future they could be added.

Reply:
I do get JSON output from `alarms listen` for every event, so that at least is working correctly - what I meant was that the JSON emitted from `alarms listen` is technically the `status-change` rather than the `alarm-notification` event. 

The `status` table is what's passed to `new_notification` (in `add_alarm_notification`), so if all of the mandatory fields for `alarm-notification` do not exist in the `status` table then the JSON output does not match the yang `alarm-notification` schema - but if I add the fields to the `status` table then the `status-change` list doesn't match the schema 😄 

Basically, `status-change` and `alarm-notification` are not the same thing but currently both rely on a single table passed in to `add_status_change`.

I'll have a think about how to work around this.

Reply:
You're right, great investigation. I think the issue can be solved if `add_alarm_notification` is passed an `alarm` and not an `status`. Could you give that a try?

Reply:
And there's an extra thing as well. Perhaps it's necessary to change `new_notification` to do a deep-copy of the `value` passed.